### PR TITLE
expression,mysql: trim the space char when comparing string or dealing  with type CHAR

### LINKED ISSUE #7085: TiDB don't trim the space char at the end of a string value.
Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.

```
create table t(a varchar(50), b varchar(50), unique a(a), index b(b));
```

2. What did you expect to see?
```
insert into t values('a ', 'a'), ('a', 'b');
```
This will raise error that 'a' have duplicate key **in mysql**.
```
insert into t values('a', 'a'), ('b', 'a     ');
select b, char_length(b) from t where b = 'a';
select b, char_length(b) from t where b = 'a   ';
```
This two select all return result as follows:
```
a 1
a 6
```

3. What did you see instead?

TiDB won't raise error in first case.
TiDB won't return the exact data, not trim the data and return two row.

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
e96cda003e07ca732d04657924dc48a0c0c4fc82

### PR DESCRIPTION
**Pull request by spongedu**
## What have you changed? (mandatory)

fix #7085 .
trim right spaces when doing query.
see https://www.percona.com/blog/2006/11/27/trailing-spaces-in-mysql/ for details.

## What is the type of the changes? (mandatory)

- Improvement (compatibility)

## How has this PR been tested? (mandatory)

Unittest

## Refer to a related PR or issue link (optional)
#7085
## Add a few positive/negative examples (optional)

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/7133)

### CODE REVIEW

File: expression/builtin_compare.go [hunk 1]
```diff
@@ -2234,7 +2235,8 @@ func compareString(args []Expression, row types.Row, ctx sessionctx.Context) (va
 	if isNull1 || err != nil {
 		return 0, isNull1, errors.Trace(err)
 	}
-	return int64(types.CompareString(arg0, arg1)), false, nil
+	// See issue #7085
```
Comment (zz-jason) [inline expression/builtin_compare.go]: how about: 
```go
// To be compatible with MySQL, we should trim the tailing space characters before
// comparing the two strings, see issue #7085 for more details.
```

File: expression/column.go [hunk 2]
```diff
@@ -236,16 +242,26 @@ func (col *Column) EvalString(ctx sessionctx.Context, row types.Row) (string, bo
 		}
 		res, err := val.ToString()
 		resLen := len([]rune(res))
-		if ctx.GetSessionVars().StmtCtx.PadCharToFullLength && col.GetType().Tp == mysql.TypeString && resLen < col.RetType.Flen {
-			res = res + strings.Repeat(" ", col.RetType.Flen-resLen)
+		if col.GetType().Tp == mysql.TypeString {
```
Comment (zz-jason) [inline expression/column.go]: how about:
```go
if col.GetType().Tp != mysql.TypeString {
	return res, err != nil, errors.Trace(err)
}

// Here we specially handle "mysql.TypeString".
if !ctx.GetSessionVars().StmtCtx.PadCharToFullLength {
	// add some comments about why we trim the tailing spaces.
	res = strings.TrimRight(res, " ")
} else if resLen < col.RetType.Flen {
	// add some comments about why we pad the tailing spaces.
	res = res + strings.Repeat(" ", col.RetType.Flen-resLen)
}
return res, err != nil, errors.Trace(err)
```

File: expression/builtin_compare.go (hunk 1 above)
Comment (spongedu) [reply expression/builtin_compare.go]: ok

File: expression/column.go (hunk 2 above)
Comment (spongedu) [reply expression/column.go]: ok

File: expression/builtin_compare.go [hunk 3]
```diff
@@ -24,6 +24,7 @@ import (
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/types/json"
 	"github.com/pingcap/tipb/go-tipb"
+	"strings"
```
Comment (jackysp) [inline expression/builtin_compare.go]: Please move this line under line 17.

File: expression/column.go [hunk 4]
```diff
@@ -72,7 +72,15 @@ func (col *CorrelatedColumn) EvalString(ctx sessionctx.Context, row types.Row) (
 	}
 	res, err := col.Data.ToString()
 	resLen := len([]rune(res))
-	if resLen < col.RetType.Flen && ctx.GetSessionVars().StmtCtx.PadCharToFullLength {
+	if col.GetType().Tp != mysql.TypeString {
+		return res, err != nil, errors.Trace(err)
+	}
+	// If sql_mode 'pad_char_to_full_length' is no set, the right trailing spaces should be removed for type 'CHAR'.
+	// See: #7085 for more details
```
Comment (jackysp) [inline expression/column.go]: It's better to add a link here.

File: expression/builtin_compare.go (hunk 3 above)
Comment (spongedu) [reply expression/builtin_compare.go]: ok

File: expression/column.go (hunk 4 above)
Comment (spongedu) [reply expression/column.go]: ok

**Comment by spongedu** [comment]
@jackysp  PTAL

File: expression/column.go [hunk 5]
```diff
@@ -73,7 +73,15 @@ func (col *CorrelatedColumn) EvalString(ctx sessionctx.Context, row chunk.Row) (
 	}
 	res, err := col.Data.ToString()
 	resLen := len([]rune(res))
-	if resLen < col.RetType.Flen && ctx.GetSessionVars().StmtCtx.PadCharToFullLength {
+	if col.GetType().Tp != mysql.TypeString {
+		return res, err != nil, errors.Trace(err)
+	}
+	// If sql_mode 'pad_char_to_full_length' is no set, the right trailing spaces should be removed for type 'CHAR'.
```
Comment (zz-jason) [inline expression/column.go]: how about: `If the 'pad_char_to_full_length' sql_mode is not set, the right padding spaces should be removed for type 'CHAR'.`

Comment (spongedu) [reply expression/column.go]: I think `If 'pad_char_to_full_length' is set in 'sql_mode' system variable, .....` would be better, that's more consistent with https://dev.mysql.com/doc/refman/5.7/en/sql-mode.html

**Comment by zz-jason** [comment]
@spongedu The following test case failed:
```sql
create table t1 (s1 binary(3));
insert into t1 values (0x61), (0x6120), (0x612020);
select hex(s1) from t1;
```

**Comment by spongedu** [comment]
@zz-jason I'll look into this.  It seems that we should distinguish type ` binary` and type `char` in this case.
```sql
mysql> create table t1 (s1 char(3));
Query OK, 0 rows affected (0.02 sec)

mysql> insert into t1 values(0x612020);
Query OK, 1 row affected (0.01 sec)

mysql> select hex(s1) from t1;
+---------+
| hex(s1) |
+---------+
| 61      |
+---------+
1 row in set (0.00 sec)

mysql> create table t2 (s1 binary(3));
Query OK, 0 rows affected (0.03 sec)

mysql> insert into t2 values(0x612020);                                                                                         
Query OK, 1 row affected (0.01 sec)

mysql> select hex(s1) from t2;
+---------+
| hex(s1) |
+---------+
| 612020  |
+---------+
1 row in set (0.00 sec)

```

It seems that

**Comment by zz-jason** [comment]
Yes, maybe we should also check the binary flag in the "FieldType".

**Comment by spongedu** [comment]
@zz-jason  PTAL

**Comment by zz-jason** [comment]
@spongedu This test failed:
```sql
create table t1 (s1 varbinary(20), s2 varbinary(20));
insert into t1 values (0x41,0x4100),(0x41,0x4120),(0x4100,0x4120);
select hex(s1), hex(s2) from t1;
select count(*) from t1 where s1 < s2;
```

**Comment by spongedu** [comment]
@zz-jason OK, I'll look into this.

**Comment by spongedu** [comment]
@zz-jason  PTAL

**Comment by zz-jason** [comment]
@spongedu This test failed:
```sql
create table t1
(
  name_id int not null auto_increment,
  name blob,
  INDEX name_idx (name(5)),
  primary key (name_id)
);

INSERT t1 VALUES(NULL,'/');
INSERT t1 VALUES(NULL,'[T,U]_axpby');

create table t2
(
  name_id int not null auto_increment,
  name char(255) binary,
  INDEX name_idx (name(5)),
  primary key (name_id)
);
INSERT t2 select * from t1;

SELECT * FROM t2 WHERE name='[T,U]_axpby';
```

**Comment by spongedu** [comment]
@zz-jason , It seems that the test base should be changed.  Result of this test with this pr is consistent with MySQL:
In MySQL:
```sql

mysql> select version();
+------------+
| version()  |
+------------+
| 5.7.17-log |
+------------+
1 row in set (0.01 sec)

mysql> create table t1
    -> (
    ->   name_id int not null auto_increment,
    ->   name blob,
    ->   INDEX name_idx (name(5)),
    ->   primary key (name_id)
    -> );
Query OK, 0 rows affected (0.03 sec)

mysql>
mysql> INSERT t1 VALUES(NULL,'/');
Query OK, 1 row affected (0.01 sec)

mysql> INSERT t1 VALUES(NULL,'[T,U]_axpby');
Query OK, 1 row affected (0.02 sec)

mysql>
mysql> create table t2
    -> (
    ->   name_id int not null auto_increment,
    ->   name char(255) binary,
    ->   INDEX name_idx (name(5)),
    ->   primary key (name_id)
    -> );
Query OK, 0 rows affected (0.03 sec)

mysql> INSERT t2 select * from t1;
Query OK, 2 rows affected (0.03 sec)
Records: 2  Duplicates: 0  Warnings: 0

mysql>
mysql> SELECT * FROM t2 WHERE name='[T,U]_axpby';
Empty set (0.00 sec)
```

With This pr:
```sql
tidb> create table t1
    -> (
    ->   name_id int not null auto_increment,
    ->   name blob,
    ->   INDEX name_idx (name(5)),
    ->   primary key (name_id)
    -> );
Query OK, 0 rows affected (0.01 sec)

tidb>
tidb> INSERT t1 VALUES(NULL,'/');
Query OK, 1 row affected (0.00 sec)

tidb> INSERT t1 VALUES(NULL,'[T,U]_axpby');
Query OK, 1 row affected (0.00 sec)

tidb>
tidb> create table t2
    -> (
    ->   name_id int not null auto_increment,
    ->   name char(255) binary,
    ->   INDEX name_idx (name(5)),
    ->   primary key (name_id)
    -> );
Query OK, 0 rows affected (0.01 sec)

tidb> INSERT t2 select * from t1;
Query OK, 2 rows affected (0.30 sec)

tidb>
tidb> SELECT * FROM t2 WHERE name='[T,U]_axpby';
Empty set (0.00 sec)

```

**Comment by zz-jason** [comment]
@spongedu It's interesting, In MySQL 5.7.22:
```sql
MySQL(localhost:3306) > SELECT * FROM t2 WHERE name='[T,U]_axpby';
+---------+-------------+
| name_id | name        |
+---------+-------------+
|       2 | [T,U]_axpby |
+---------+-------------+
1 row in set (0.00 sec)

MySQL(localhost:3306) > select version();
+-----------+
| version() |
+-----------+
| 5.7.22    |
+-----------+
1 row in set (0.00 sec)
```

**Comment by spongedu** [comment]
@zz-jason  It seems that we can't distinguish `char(100) binary` type and `binary(100)` type at runtime.  They both have type `TypeString`, set `BinaryFlag`, with `binary` charset and collate.  This case fails here. MySQL trims right trailing spaces for type `char` no matter whether `binary` flag is set on it.

**Comment by spongedu** [comment]
@zz-jason  These cases fail due to `mocktikv` fail to recover `FieldType` correctly (missing `flag`, `charset`, `collate`, `flen` etc) from tipb. We should check if `coprocess` deal with it correctly.

**Comment by zz-jason** [comment]
Yes, I think we'd better use another PR to solve this issue, and cherrypick it to release-2.0

**Comment by zz-jason** [comment]
> @zz-jason These cases fail due to `mocktikv` fail to recover `FieldType` correctly (missing `flag`, `charset`, `collate`, `flen` etc) from tipb. We should check if `coprocess` deal with it correctly.

Do we still have this issue? @XuHuaiyu @yangxuanjia @breeswish PTAL

**Comment by zz-jason** [comment]
@spongedu This PR is conflicted with the master branch, I added a `status/ReqChange` label on it. I'll remove it when the conflict is resolved and it's ready to be reviewed.

**Comment by zz-jason** [comment]
I'm going to close this PR due to outdated, fell free to reopen it if you are going to process it again in the future.

**Pull request closed without merging**